File: backend/mystaticfiles/javascript/game/team.js
import * as dc from "./client_define.js"
import * as d from "./define.js"
import * as paddle from "./paddle.js"
import * as utils from "./pg_utils.js"
// import * as Math from 'math.js'


export class Team {
	constructor ( numberOfPlayers, team){
		if (numberOfPlayers < 1)
			numberOfPlayers = 1
		else if (numberOfPlayers > d.TEAM_MAX_PLAYER)
			numberOfPlayers = d.TEAM_MAX_PLAYER

		this.team = team

		let xPos = 0
		if (this.team == d.TEAM_LEFT)
			xPos = dc.AREA_RECT[0] + d.AREA_BORDER_SIZE * 2
		else
			xPos = dc.AREA_RECT[0] + dc.AREA_RECT[2] - d.AREA_BORDER_SIZE * 2

		this.paddles = []
		if (numberOfPlayers == 1)
		{
			console.log("team create paddle at : [ " + xPos + " ] [ " + dc.AREA_RECT[1] + " + " + dc.AREA_RECT[3] + " / 2  = " + dc.AREA_RECT[1] + dc.AREA_RECT[3] / 2 + " ]")
			this.paddles.push( new paddle.Paddle(xPos, dc.AREA_RECT[1] + dc.AREA_RECT[3] / 2, 0, this.team))
		}
		else
		{
			this.paddles.push( new paddle.Paddle(xPos, dc.AREA_RECT[1] + Math.floor(dc.AREA_RECT[3] / 3), 0, this.team))
			this.paddles.push( new paddle.Paddle(xPos, dc.AREA_RECT[1] + Math.floor(dc.AREA_RECT[3] / 3) * 2, 1, this.team))
		}

		this.score = 0
		// list of power up who try to use : [power up id, paddle id, power up used (bool)]
		this.powerUpTryUse = []
	}


	tick( delta, paddlesKeyState, updateTime){
		// Check input
		for (let i = 0; i < this.paddles.length; i++) 
		{
			if (updateTime)
				this.paddles[i].updateTimes(delta)

			let keyId = i
			if (this.team == d.TEAM_RIGHT)
				keyId += d.TEAM_MAX_PLAYER

			if (paddlesKeyState[keyId * 4 + d.KEY_UP])
				this.paddles[i].move("up", delta)
			if (paddlesKeyState[keyId * 4 + d.KEY_DOWN])
				this.paddles[i].move("down", delta)
		}
	}

	draw( win, powerUpEnable)
	{
		for (const p of this.paddles)
			p.draw(win)

		if (this.team == dc.TEAM_LEFT)
		{			
			utils.drawText(win, "SCORE : " + str(this.score), (75, 75 / 2), (255, 255, 255), 30, null, "mid-left")
			if (powerUpEnable)
			{
				utils.drawText(win, this.paddles[0].powerUp, (50, 70), (255, 255, 255), 30, null, "mid-right")
				if (this.paddles.length == 2)
				utils.drawText(win, this.paddles[1].powerUp, (50, dc.WIN_HEIGHT - 70), (255, 255, 255), 30, null, "mid-right")
			}
		}

		else
		{
			utils.drawText(win, "SCORE : " + this.score, (dc.WIN_WIDTH - 75, 75 / 2), (255, 255, 255), 30, null, "mid-right")
			if (powerUpEnable)
			{
				utils.drawText(win, this.paddles[0].powerUp, (dc.WIN_WIDTH - 50, 70), (255, 255, 255), 30, null, "mid-left")
				if (this.paddles.length == 2)
				utils.drawText(win, this.paddles[1].powerUp, (dc.WIN_WIDTH - 50, dc.WIN_HEIGHT - 70), (255, 255, 255), 30, null, "mid-left")
			}
		}
	}


	applyPowerUpToPaddles( powerUp)
	{
		for (const p of this.paddles)
			p.addPowerUpEffect(powerUp)
	}
}

File: backend/pong_server/server_side/team.py
from define import *

import server_side.paddle as paddle
import random

class Team:
	def __init__(self, numberOfPlayers:int, team:int) -> None:
		if numberOfPlayers < 1:
			numberOfPlayers = 1
		elif numberOfPlayers > TEAM_MAX_PLAYER:
			numberOfPlayers = TEAM_MAX_PLAYER

		self.team = team

		if self.team == TEAM_LEFT:
			xPos = AREA_BORDER_SIZE * 2
		else:
			xPos = AREA_SIZE[0] - AREA_BORDER_SIZE * 2

		self.paddles = []
		if numberOfPlayers == 1:
			self.paddles.append(paddle.Paddle(xPos, AREA_SIZE[1] // 2 + random.randint(-50, 50), 0, self.team))
		else:
			self.paddles.append(paddle.Paddle(xPos, AREA_SIZE[1] // 3 + random.randint(-50, 50), 0, self.team))
			self.paddles.append(paddle.Paddle(xPos, AREA_SIZE[1] // 3 * 2 + random.randint(-50, 50), 1, self.team))

		self.score = 0
		# list of power up who try to use : [power up id, paddle id, power up used (bool)]
		self.powerUpTryUse = []


	def tick(self, delta:float, paddlesKeyState:list, updateTime:bool) -> None:
		# Check power up try to used
		for powerUp in self.powerUpTryUse:
			# if a power up is user, remove it to the player
			if powerUp[2]:
				if powerUp[0] == POWER_UP_BALL_FAST:
					self.paddles[powerUp[1]].powerUpInCharge.append(POWER_UP_BALL_FAST)
				elif powerUp[0] == POWER_UP_BALL_WAVE:
					self.paddles[powerUp[1]].powerUpInCharge.append(POWER_UP_BALL_WAVE)
				elif powerUp[0] == POWER_UP_BALL_INVISIBLE and self.paddles[powerUp[1]].waitUsePowerUp == 0:
					self.paddles[powerUp[1]].powerUpInCharge.append(POWER_UP_BALL_INVISIBLE)

				self.paddles[powerUp[1]].waitUsePowerUp = POWER_UP_USE_COOLDOWN
				self.paddles[powerUp[1]].powerUp = POWER_UP_NONE

		self.powerUpTryUse.clear()


		# Check input
		for i in range(len(self.paddles)):
			if updateTime:
				self.paddles[i].updateTimes(delta)

			keyId = i
			if self.team == TEAM_RIGHT:
				keyId += TEAM_MAX_PLAYER

			if paddlesKeyState[keyId * 4 + KEY_UP]:
				self.paddles[i].move("up", delta)
			if paddlesKeyState[keyId * 4 + KEY_DOWN]:
				self.paddles[i].move("down", delta)
			if paddlesKeyState[keyId * 4 + KEY_POWER_UP]:
				powerUp = self.paddles[i].powerUp
				if powerUp != POWER_UP_NONE:
					self.powerUpTryUse.append([powerUp, i, False])


	def applyPowerUpToPaddles(self, powerUp):
		for p in self.paddles:
			p.addPowerUpEffect(powerUp)
